hivemind/p2p/p2p_daemon_bindings/control.py

```diff
@@ -93,6 +103,99 @@ async def listen(self) -> AsyncIterator["ControlClient"]:
         async with server:
             yield self
 
+    async def _read_from_persistent_conn(self, reader: asyncio.StreamReader):
+        while True:
+            resp: p2pd_pb.Response = p2pd_pb.Response()  # type: ignore
+            await read_pbmsg_safe(reader, resp)
+
+            if resp.HasField("callUnaryResponse"):
+                call_id = uuid.UUID(bytes=resp.callUnaryResponse.callId)
+
+                if call_id in self.pending_calls and resp.callUnaryResponse.HasField("result"):
+                    self.pending_calls[call_id].set_result(resp.callUnaryResponse.result)
+                elif call_id in self.pending_calls and resp.callUnaryResponse.HasField("error"):
+                    remote_exc = RemoteException(str(resp.callUnaryResponse.error))
+                    self.pending_calls[call_id].set_exception(remote_exc)
+                else:
+                    logger.debug(f"received unexpected unary call")
+
+            elif resp.HasField("requestHandling"):
+                asyncio.create_task(self._handle_persistent_request(resp.requestHandling))
+                pass
+
+    async def _write_to_persistent_conn(self, writer: asyncio.StreamWriter):
+        while True:
+            msg = await self.pending_messages.get()
+            await write_pbmsg(writer, msg)
+
+    async def _handle_persistent_request(self, request):
+        assert request.proto in self.unary_handlers
+
+        try:
+            response_payload: bytes = self.unary_handlers[request.proto](request.data)
+            response = p2pd_pb.CallUnaryResponse(callId=request.callId, result=response_payload)
+        except Exception as e:
+            response = p2pd_pb.CallUnaryResponse(callId=request.callId, error=repr(e))
+
+        await self.pending_messages.put(
+            p2pd_pb.Request(type=p2pd_pb.Request.SEND_RESPONSE_TO_REMOTE, sendResponseToRemote=response)
+        )
+
+    async def _handler(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
+        pb_stream_info = p2pd_pb.StreamInfo()  # type: ignore
+        await read_pbmsg_safe(reader, pb_stream_info)
+        stream_info = StreamInfo.from_protobuf(pb_stream_info)
+        try:
+            handler = self.handlers[stream_info.proto]
+        except KeyError as e:
+            # should never enter here... daemon should reject the stream for us.
+            writer.close()
+            raise DispatchFailure(e)
+        await handler(stream_info, reader, writer)
+
+    async def _ensure_persistent_conn(self):
+        if not self._pers_conn_open:
+            reader, writer = await self.daemon_connector.open_persistent_connection()
+            asyncio.create_task(self._read_from_persistent_conn(reader))
+            asyncio.create_task(self._write_to_persistent_conn(writer))
```

Comment:
Let's save the tasks to the class fields and cancel them in the end of `listen()`.

Also, let's wrap the code in `_write_to_persistent_conn` with the `with closing(writer):` statement.